### PR TITLE
Add support for app.import(…, {outputFile: …})

### PR DESCRIPTION
**Pull request by stefanpenner**
[rendered](https://github.com/ember-cli/rfcs/blob/app-import-output-file/active/0000-app-import-output-file.md)

### CODE REVIEW

**Comment by rwjblue** [comment]
@stefanpenner - I'm :+1: on this.  I think additional hooks would be fine, but it seems additive and can be done after this lands (IMHO).

**Comment by Gaurav0** [comment]
I'm in favor of this.

You should make clear if this applies to CSS files or not.

**Comment by stefanpenner** [comment]
> You should make clear if this applies to CSS files or not.

is ordered by import order based concatenation alright ?

**Comment by rwjblue** [comment]
@stefanpenner - Yes, I think import order is the best we can do for CSS.

**Comment by stefanpenner** [comment]
I'll call out explicitly, that it will blindly concat ANY files in import order. video/image/html/javascript/css etc.etc.

**Comment by rwjblue** [comment]
Hehe, binary file concat is both a bug and a feature depending on what you want :smiling_imp:

**Comment by stefanpenner** [comment]
> Hehe, binary file concat is both a bug and a feature depending on what you want

c

**Comment by stefanpenner** [comment]
> You should make clear if this applies to CSS files or not.
> is ordered by import order based concatenation alright

updated

**Comment by Gaurav0** [comment]
concatenated is misspelled.

**Comment by stefanpenner** [comment]
> concatenated is misspelled.

can you provide line-comments, easier to track down which is mispelled

File: active/0000-app-import-output-file.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+It is also common for developers to want to group various dependencies
+together, and then lazy-load them in the routes they are required.
+
+Although not as automatic as we would like, it does provide a rather elegant
+escape valve. Further work will likely continue to explore automation.
+
+# Detailed design
+
+`outputFile:` option, specifies the target file for the given import. If
+multiple imports share an outputFile, they will be concattenated (regardless of
```
Comment (Gaurav0) [inline active/0000-app-import-output-file.md]: concatenated is misspelled

**Comment by searls** [comment]
I'm a huge :+1: on this. I've seen two teams have to make concessions about how they bundle & distribute assets to fit the existing model, and this would have definitely helped in at least one case.

File: active/0000-app-import-output-file.md (hunk 1 above)
Comment (stefanpenner) [inline active/0000-app-import-output-file.md]: @Gaurav0 thank you kindly sir, updated (and hopefully fixed)

**Comment by stefanpenner** [comment]
@patocallaghan / @searls either of you interested in implementing this?

File: active/0000-app-import-output-file.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+complementary.
+
+Dropping down to broccoli is a solution available today, but for this problem,
+it feels like a slightly too low level of abstraction.
+
+# Unresolved questions
+
+* how does this relate to `type` in `app.import(..., { type: ... })` ?
+* should additional build-steps be allowed for specific output files? (I
+  suspect maybe, but a future RFC can likely explore) What parts of the design
```
Comment (Gaurav0) [inline active/0000-app-import-output-file.md]: Unfinished question

**Comment by patocallaghan** [comment]
@stefanpenner I'd be interested in taking this on once it syncs up with other stuff I have on. What sort of timeline were you thinking?

**Comment by stefanpenner** [comment]
Just whenever someone has time. As usually sooner rather then later. But afew days to a week or so would be ideal.

**Comment by patocallaghan** [comment]
I won't be able to commit until late next week so if someone wants to take it on beforehand, fine by me..

**Comment by cibernox** [comment]
This interesting for some work I'd like to do with web workers. However I think that the addon ecosystem should be able to use this.

By example, imagine that ember data evolves to perform ajax requests and payload normalization in a web-worker. The script that this addon would generate (p.e `ember-data-worker.js`) can:

A) Contain Ember (only parts?) + Ember-data + User-defined serializers/adapters
B) Contain only User-defined serializers/adapters but use `importScripts` to load a secondary `ember-data-worker/vendor.js`) that contains Ember + Ember Data + any other dependency. 

That makes me think that addons should have the ability to add their files to a specific tree. This RFC covers part of this scenario, but not all.

**Comment by stefanpenner** [comment]
Addons(who are direct descendants of the app) can already run app.import

**Comment by yaymukund** [comment]
:+1: this would be useful for quickly integrating web workers

**Comment by stefanpenner** [comment]
@rwjblue can you :eyes: , want to make sure i haven't missed anything.

**Comment by rwjblue** [comment]
@stefanpenner - This is good to go from my perspective.  Now we need someone to implement :smiling_imp:...

**Comment by MiguelMadero** [comment]
@rwjblue @stefanpenner do you still need someone to implement this? I'm working on something similar and I have some time for it. Just wanted to avoid duplicating efforts in case someone is already tackling it. Please let me know either way.

**Comment by rwjblue** [comment]
@MiguelMadero - Nope, it is included in ember-cli@2.4.2

**Comment by MiguelMadero** [comment]
nice! Thanks

**Comment by MiguelMadero** [comment]
Is there a reason why addons aren't supported for this? 
Seems like all of the addon code goes into addons.js and later merged into vendor.js, but there isn't an option to concatenate them. I'm thinking that we could do

```
app.import('addons/ember-power-select', {outputFile: 'vendor-deferred.js'});
```

Does that sound reasonable?

**Comment by stefanpenner** [comment]
that API isn't unreasonable, but it is a departure from what we do today. I think its worth exploring though, do you mind fleshing this out more in an RFC ?

**Comment by MiguelMadero** [comment]
Yes. I think it's basically an extension to this RFC. Let me draft something and we can discus it there. I might need some help around the drawbacks and the potential overlap with ember-packager.

**Comment by stefanpenner** [comment]
@MiguelMadero sounds good, ultimately the packager spike didn't really have a public API. Merely internal re-construction, to enable. So the high level API is fair-game.

**Pull request merged**